// control-errores/esm2020/lib/control-errores.module.mjs
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import * as i0 from "@angular/core";
export class ControlErroresModule {
}
ControlErroresModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: ControlErroresModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ControlErroresModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.12", ngImport: i0, type: ControlErroresModule, imports: [HttpClientModule] });
ControlErroresModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: ControlErroresModule, imports: [HttpClientModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: ControlErroresModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [HttpClientModule],
                    exports: [],
                }]
        }] });
